### ISSUE TITLE
Front end infers dynamic for ".then" invocations

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#33044]
Consider the following code:

```dart
void test(Future<int> f) {
  var t1 = f.then((int x) {});
  var t2 = f.then((int x) { return; });
  var t3 = f.then((int x) { return null; });
  var t4 = f.then((int x) => null);
  var t5 = f.then((int x) { return x.toString(); });
  var t6 = f.then((int x) => x.toString());
  var t7 = f.then((x) {});
  var t8 = f.then((x) { return; });
  var t9 = f.then((x) { return null; });
  var t10 = f.then((x) => null);
  var t11 = f.then((x) { return x.toString(); });
  var t12 = f.then((x) => x.toString());
}
```

The analyzer infers `t5`, `t6`, `t11`, and `t12` to have type `Future<String>`, and the others to have type `Future<Null>`.  The front end infers the type of `t1` through `t12` to be `dynamic`.

This has user-visible consequences.  For example, the analyzer considers this to be an error, and the front end doesn't:

```dart
void test(Future<int> f) async {
  var t = f.then((int x) { return x.toString(); });
  print(t.foo);
}
```

I believe the analyzer behavior is correct.  @leafpetersen can you confirm?

**Comment by leafpetersen** [comment]
I suspect that all of the `.toString` examples are instances of this bug: https://github.com/dart-lang/sdk/issues/32414 .

For the remainder: `Future<dynamic>` is always wrong.  For t1, t2, t7, and t8 the analyzer is still implementing legacy behavior: we should infer `Future<void>` there, and I will be working on that change as soon as a fix for return warnings has rolled internally (unblocking the fixes for this).

For the rest, `Future<Null>` is correct.

**Comment by leafpetersen** [comment]
https://github.com/dart-lang/sdk/issues/32843 is the issue for void inference.

**Comment by stereotype441** [comment]
> I suspect that all of the `.toString` examples are instances of this bug: https://github.com/dart-lang/sdk/issues/32414 .

@leafpetersen in that case I would have expected the front end to infer a type of `Future<dynamic>`.  But it's inferring a type of `dynamic`.

**Comment by leafpetersen** [comment]
sorry, I didn't read closely enough.  I can't reproduce this though: the code below, run with dart --preview-dart-2 from fairly recent bleeding edge, has the expected errors (I think it stops reporting errors after some limit, but if you delete the first 5, you get the next 5 as expected).  Am I still missing something?

```
import 'dart:async';

void test(Future<int> f) {
  int x;
  var t1 = f.then((int x) {});
  x = t1;
  t1.foo;
  var t2 = f.then((int x) { return; });
  x = t2;
  t2.foo;
  var t3 = f.then((int x) { return null; });
  x = t3;
  t3.foo;
  var t4 = f.then((int x) => null);
  x = t4;
  t4.foo;
  var t5 = f.then((int x) { return x.toString(); });
  x = t5;
  t5.foo;
  var t6 = f.then((int x) => x.toString());
  x = t6;
  t6.foo;
  var t7 = f.then((x) {});
  x = t7;
  t7.foo;
  var t8 = f.then((x) { return; });
  x = t8;
  t8.foo;
  var t9 = f.then((x) { return null; });
  x = t9;
  t9.foo;
  var t10 = f.then((x) => null);
  x = t10;
  t10.foo;
  var t11 = f.then((x) { return x.toString(); });
  x = t11;
  t11.foo;
  var t12 = f.then((x) => x.toString());
  x = t12;
  t12.foo;
}

void main() {
  test(new Future.value(3));
}
```

**Comment by stereotype441** [comment]
@leafpetersen Oops, you are right.  I forgot to include `import 'dart:async';` in my test script (and failed to notice the error message the front end issued).

When I include the correct import, I get the same results as you.  Since the remaining problems are already covered by open bugs, I'm closing this bug.
